Feat lesson27

yuka830 opened the pull request:
# [Lesson27](https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#27)
```
index.htmlページ遷移時に 学習の為のローカルストレージのトークンを参照して 
もし {token: far0fja*ff]afaawfqrlzkfq@aq9283af}の値があれば 作ったコンテンツUI画面を
何もない場合はログイン画面(login.html)に遷移するように作ってください
```
## [CodeSandbox](https://codesandbox.io/s/lesson27-hb7y18?file=/index.html)
Last commit: [18c0af7](https://github.com/yuka830/js-lessons/commit/18c0af79ad2f94b6073d97fa09d6b5151c96c9af)

I'm not sure I understand this lesson's meaning correctly. Please kindly check my code. Thanks.

kenmori commented:
> 何もない場合はログイン画面(login.html)に遷移するように作ってください

見たところ
1. ログインしてから
2. コンテンツ画面でローカルストレージを削除
3. その画面にとどまる

3でログインページにいかないのですよね..

443n0511 commented on lesson27/js/token.js:
```diff
@@ -0,0 +1,3 @@
+localStorage.getItem("token")
+  ? (window.location.href = "./news.html")
+  : (window.location.href = "./login.html");
```
IMO
It will be limited.
`(window.location.href = "./news.html")`
I feel like I don't need this!

yuka830 replied:
Thank you!!
If I implement like below, it will stop working when I open index.html if it has token.
`if (!localStorage.getItem("token")) window.location.href = "./login.html";`

I think I might misunderstand something...
Is this `index.html` mentioned in this lesson perhaps referring to the news content?
I had created a new index.html separate from the news.html💦 

↑I'm sorry if I'm saying something completely different.

443n0511 replied:
Yes, the contents of index.html should contain newscontent!
[codesandbox](https://codesandbox.io/s/lesson27-forked-e1irqw?file=/login.html)
This way, even if you intentionally delete the local storage after logging in, you won't be redirected to the content page.

yuka830 replied:
I appreciate your kind instruction. I've fixed about this.
[633274e](https://github.com/yuka830/js-lessons/pull/32/commits/633274e4288e53c83d84a234e788959dadda707e)
I deleted `token.js` because the news page shows up a moment before moving to the login page when it doesn't have token. I implemented this code in `index.html`'s script. 

> [9739583](https://github.com/yuka830/js-lessons/pull/32/commits/9739583076248e484301544daf9cb8c2c73ace62) ←This part has been removed from this sandbox.

Oh, I think I might change something when you were checking my CodeSandbox.
Sorry about that.

yuka830 commented:
@kenmori 
> 何もない場合はログイン画面(login.html)に遷移するように作ってください
> 
> 見たところ
> 
> 1. ログインしてから
> 2. コンテンツ画面でローカルストレージを削除
> 3. その画面にとどまる
> 
> 3でログインページにいかないのですよね..

I added `addEventLister` and to monitor the token in newsUi.js.
[9739583](https://github.com/yuka830/js-lessons/pull/32/commits/9739583076248e484301544daf9cb8c2c73ace62)

yuka830 commented:
@kenmori @443n0511 
What I've changed ↓
- Delete `index.html` and rename `news.html` to `index.html`
- Implement only the pattern that it doesn't have token in `token.js`.  But the news page shows up a moment before moving to the login page when it’s in this pattern. So call the js code in `index.html` instead of using `token.js`, and delete `token.js`.
- Monitoring when token is deleted in the Application panel of DevTools with `addEventListener(storage)`

Last commit:[9d58265](https://github.com/yuka830/js-lessons/pull/32/commits/9d58265ce210a0ecee028d010e6827bd2d16b368)
[CodeSandbox](https://codesandbox.io/s/lesson27-hb7y18?file=/index.html)

ghost commented:
MUST
- login.html 11 line items
I guess I'd better fix this too.
If you go directly to login.html after a successful login, you will be redirected to news.html
```html
  <script>localStorage.getItem("token") && (window.location.href = "./news.html");
  </script>
```
<img width="821" alt="スクリーンショット 2022-02-21 14 04 06" src="https://user-images.githubusercontent.com/43788628/154892741-75189963-5f0b-45f9-b80a-a662b90e0097.png">

yuka830 commented:
@443n0511 
> login.html 11 line items
I guess I'd better fix this too.
If you go directly to login.html after a successful login, you will be redirected to news.html

I'm sorry about that, I totally forgot to fix this.
Thank you for your help!!
[3dd7ebe](https://github.com/yuka830/js-lessons/pull/32/commits/3dd7ebe4f340cd58a4f9b12ab6350d0d9df47e4a)

yuka830 commented:
## Feat

Added the animation for the input. When error messages show up, make input forms shaking.
( register.html & login.html)
[31f387a](https://github.com/yuka830/js-lessons/pull/32/commits/31f387a5ca55e51c7a331ada08b683af08c5f156)
[7739e3b](https://github.com/yuka830/js-lessons/pull/32/commits/7739e3bd8b746afec8e7684f8668e573d39d1a3e)
Added a modal which shows loading and login result after clicking the login button.
[07b5115](https://github.com/yuka830/js-lessons/pull/32/commits/07b5115de8d57d76ad5b764f39a87a5248890f10)

https://user-images.githubusercontent.com/66966699/155680197-7943b582-6223-47bb-aa8b-40ff8ac3f0e9.gif

kenmori commented on lesson27/js/login.js:
```diff
@@ -28,23 +28,30 @@ const loginHandler = async (e) => {
   } catch {
     result = false;
   } finally {
+    changeModalImgAndText(result);
```
ここなのですが、

```js
  try {
    const token = await login(inputVal);
  } catch {
    result = false;
  } finally {
    changeModalImgAndText(result);
    changeLocation(result);
```
    
    
こうするといいかもです
    
```js
      try {
    const token = await login(inputVal);
  } catch(e) {
    changeModalImgAndErrorText(e);
    changeErrorLocation();
    return
  }
    changeModalImgAndText(result);
    changeLocation(result);
```
    
利点は
    finallyはあくまで共通の後処理みたいなイメージで、
    その中でerrorだったらこうするみたいな実装はしないといいです
    なぜならエラーのことはcatchの中でできるはずなのでそこに書くとif文も無くなって明確です
    こういう場合関数は別で用意したいです
    またcatchでもらえるイベントも使えそうです

yuka830 replied:
Thank you for teaching me. I've fixed :)

[Question]
I'm sorry but I don't understand about `changeModalImgAndErrorText(e);`.
Here `e` is `{ ok: false, code: 401 }`, how do I use this in this function?
Or you meant we can use it for the error message in console?

For now I tried to implement without `e`. 
[7888f49](https://github.com/yuka830/js-lessons/pull/32/commits/7888f4914fb326e5a58f6a4d3bac18a704aa5979)
